satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,312 @@
+	if criteria.DistinctIP {
+		reputableNodeQuery = assemble(partialQuery{selection: reputableNodeSelection, condition: reputableNodesCondition, orderBy: "last_net", limit: reputableNodeCount})
+		reputableNodeQuery = selectRandomFrom(reputableNodeQuery, "filteredcandidates", reputableNodeCount)
+	}
+
+	query := union(newNodeQuery, reputableNodeQuery)
+
+	rows, err := cache.db.Query(ctx, cache.db.Rebind(query.query), query.args...)
+	if err != nil {
+		return nil, nil, Error.Wrap(err)
+	}
+	defer func() { err = errs.Combine(err, rows.Close()) }()
+
+	for rows.Next() {
+		var node overlay.SelectedNode
+		node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+		var lastIPPort sql.NullString
+		var isNew bool
+		err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort, &isNew)
+		if err != nil {
+			return nil, nil, err
+		}
+		if lastIPPort.Valid {
+			node.LastIPPort = lastIPPort.String
+		}
+		if isNew {
+			newNodes = append(newNodes, &node)
+		} else {
+			reputableNodes = append(reputableNodes, &node)
+		}
+		if len(newNodes)+len(reputableNodes) == reputableNodeCount+newNodeCount {
+			break
+		}
+	}
+	return reputableNodes, newNodes, nil
+}
+
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string, isNewNodeQuery bool) (condition, error) {
+	var conds conditions
+	if isNewNodeQuery {
+		conds.add(
+			"total_audit_count < ?",
+			criteria.AuditCount,
+		)
+	} else {
+		conds.add(
+			"total_audit_count >= ? AND total_uptime_count >= ?",
+			criteria.AuditCount, criteria.UptimeCount,
+		)
+	}
+	conds.add(`disqualified is null
+			AND suspended is null
+			AND exit_initiated_at is null`)
+
+	conds.add("type = ?", int(pb.NodeType_STORAGE))
+	conds.add("free_disk >= ?", criteria.FreeDisk)
+	conds.add("last_contact_success > ?", time.Now().Add(-criteria.OnlineWindow))
+
+	if criteria.MinimumVersion != "" {
+		v, err := version.NewSemVer(criteria.MinimumVersion)
+		if err != nil {
+			return condition{}, Error.New("invalid node selection criteria version: %v", err)
+		}
+		conds.add(
+			"(major > ? OR (major = ? AND (minor > ? OR (minor = ? AND patch >= ?)))) AND release ",
+			v.Major, v.Major, v.Minor, v.Minor, v.Patch,
+		)
+	}
+	cond := combine(conds...)
+
+	if len(excludedIDs) > 0 {
+		excludedIDsQuery := " AND id NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+		cond.addQuery(excludedIDsQuery)
+		for _, id := range excludedIDs {
+			cond.addArg(id)
+		}
+	}
+	if criteria.DistinctIP {
+		if len(excludedNetworks) > 0 {
+			excludedNetworksQuery := " AND last_net NOT IN (?" + strings.Repeat(", ?", len(excludedIDs)-1) + ") "
+			cond.addQuery(excludedNetworksQuery)
+			for _, subnet := range excludedNetworks {
+				cond.addArg(subnet)
+			}
+		}
+		cond.addQuery("AND last_net <> ''")
```

Comment:
add whitespace in the beginning of this query

Reply:
ok added